// @nx/vite/src/utils/e2e-web-server-info-utils.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    getReactRouterE2EWebServerInfo: function() {
        return getReactRouterE2EWebServerInfo;
    },
    getViteE2EWebServerInfo: function() {
        return getViteE2EWebServerInfo;
    }
});
const _devkit = require("@nx/devkit");
const _e2ewebserverinfoutils = require("@nx/devkit/src/generators/e2e-web-server-info-utils");
async function getViteE2EWebServerInfo(tree, projectName, configFilePath, isPluginBeingAdded, e2ePortOverride) {
    var _nxJson_targetDefaults, _nxJson_targetDefaults_dev_options, _nxJson_targetDefaults1, _nxJson_targetDefaults2, _nxJson_targetDefaults_serve_options, _nxJson_targetDefaults3;
    const nxJson = (0, _devkit.readNxJson)(tree);
    let e2ePort = e2ePortOverride != null ? e2ePortOverride : 4200;
    if (((_nxJson_targetDefaults = nxJson.targetDefaults) == null ? void 0 : _nxJson_targetDefaults['dev']) && ((_nxJson_targetDefaults1 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_dev_options = _nxJson_targetDefaults1['dev'].options) == null ? void 0 : _nxJson_targetDefaults_dev_options.port) || ((_nxJson_targetDefaults2 = nxJson.targetDefaults) == null ? void 0 : _nxJson_targetDefaults2['serve']) && ((_nxJson_targetDefaults3 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_serve_options = _nxJson_targetDefaults3['serve'].options) == null ? void 0 : _nxJson_targetDefaults_serve_options.port)) {
        var _nxJson_targetDefaults_dev_options1, _nxJson_targetDefaults4, _nxJson_targetDefaults_serve_options1, _nxJson_targetDefaults5;
        var _nxJson_targetDefaults_dev_options_port;
        e2ePort = (_nxJson_targetDefaults_dev_options_port = (_nxJson_targetDefaults4 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_dev_options1 = _nxJson_targetDefaults4['dev'].options) == null ? void 0 : _nxJson_targetDefaults_dev_options1.port) != null ? _nxJson_targetDefaults_dev_options_port : (_nxJson_targetDefaults5 = nxJson.targetDefaults) == null ? void 0 : (_nxJson_targetDefaults_serve_options1 = _nxJson_targetDefaults5['serve'].options) == null ? void 0 : _nxJson_targetDefaults_serve_options1.port;
    }
    return (0, _e2ewebserverinfoutils.getE2EWebServerInfo)(tree, projectName, {
        plugin: '@nx/vite/plugin',
        serveTargetName: 'devTargetName',
        serveStaticTargetName: 'previewTargetName',
        configFilePath
    }, {
        defaultServeTargetName: 'dev',
        defaultServeStaticTargetName: 'preview',
        defaultE2EWebServerAddress: `http://localhost:${e2ePort}`,
        defaultE2ECiBaseUrl: 'http://localhost:4300',
        defaultE2EPort: e2ePort
    }, isPluginBeingAdded);
}
async function getReactRouterE2EWebServerInfo(tree, projectName, configFilePath, isPluginBeingAdded, e2ePortOverride) {
    var _ref;
    const e2ePort = (_ref = e2ePortOverride != null ? e2ePortOverride : parseInt(process.env.PORT)) != null ? _ref : 4200;
    return await (0, _e2ewebserverinfoutils.getE2EWebServerInfo)(tree, projectName, {
        plugin: '@nx/react/router-plugin',
        serveTargetName: 'devTargetName',
        serveStaticTargetName: 'devTargetName',
        configFilePath
    }, {
        defaultServeTargetName: 'dev',
        defaultServeStaticTargetName: 'dev',
        defaultE2EWebServerAddress: `http://localhost:${e2ePort}`,
        defaultE2ECiBaseUrl: 'http://localhost:4200',
        defaultE2EPort: e2ePort
    }, isPluginBeingAdded);
}
