// @nx/vite/src/utils/versions.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    analogVitestAngular: function() {
        return analogVitestAngular;
    },
    edgeRuntimeVmVersion: function() {
        return edgeRuntimeVmVersion;
    },
    happyDomVersion: function() {
        return happyDomVersion;
    },
    jitiVersion: function() {
        return jitiVersion;
    },
    jsdomVersion: function() {
        return jsdomVersion;
    },
    nxVersion: function() {
        return nxVersion;
    },
    vitePluginDtsVersion: function() {
        return vitePluginDtsVersion;
    },
    vitePluginReactSwcVersion: function() {
        return vitePluginReactSwcVersion;
    },
    vitePluginReactVersion: function() {
        return vitePluginReactVersion;
    },
    viteV5Version: function() {
        return viteV5Version;
    },
    viteVersion: function() {
        return viteVersion;
    },
    vitestCoverageIstanbulVersion: function() {
        return vitestCoverageIstanbulVersion;
    },
    vitestCoverageV8Version: function() {
        return vitestCoverageV8Version;
    },
    vitestV1CoverageIstanbulVersion: function() {
        return vitestV1CoverageIstanbulVersion;
    },
    vitestV1CoverageV8Version: function() {
        return vitestV1CoverageV8Version;
    },
    vitestV1Version: function() {
        return vitestV1Version;
    },
    vitestVersion: function() {
        return vitestVersion;
    }
});
const nxVersion = require('../../package.json').version;
const viteVersion = '^6.0.0';
const viteV5Version = '^5.0.0';
const vitestVersion = '^3.0.0';
const vitestV1Version = '^1.3.1';
const vitePluginReactVersion = '^4.2.0';
const vitePluginReactSwcVersion = '^3.5.0';
const jsdomVersion = '~22.1.0';
const vitePluginDtsVersion = '~4.5.0';
const happyDomVersion = '~9.20.3';
const edgeRuntimeVmVersion = '~3.0.2';
const jitiVersion = '2.4.2';
const analogVitestAngular = '~1.14.1';
const vitestCoverageV8Version = '^3.0.5';
const vitestV1CoverageV8Version = '^1.0.4';
const vitestCoverageIstanbulVersion = '^3.0.5';
const vitestV1CoverageIstanbulVersion = '^1.0.4';
